cpp/src/datetime/datetime_ops.cu

```diff
@@ -116,6 +116,35 @@ struct ceil_timestamp {
   }
 };
 
+template <datetime_component COMPONENT>
```

Comment:
I wrote an example of how to do this dispatch here: https://godbolt.org/z/YaPrqdWjY

I'm meeting on Monday with @skirui-source and @mayankanand007 to explain how this works. We may choose to pursue this in a future PR, especially if we add support for "round" (which behaves similarly to ceil and floor) but it is out of scope for this PR.

Reply:
Can we make the COMPONENT an argument to `operator()` instead of a template parameter for the functor? It should be trivial to resolve at runtime and avoids instantiating a new template operator for every single component.

Reply:
Looking at how this is called with `thrust::transform`, it might be cleaner to make this an argument to the functor constructor instead of `operator()`. I agree it does not need to be a template parameter.

Reply:
Yeah that would be fine too and would require the least change to the existing code since the component would just move from the template args list into the brace initializer when constructed for type dispatch.

Reply:
Updated code example from discussion today with @skirui-source / @mayankanand007: https://godbolt.org/z/qo43ofTff